src/NuGet.Clients/NuGet.Tools/PMUIStarterDelegateService.cs

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Telemetry;
+
+namespace NuGetVSExtension
+{
+    [Export(typeof(IPMUIStarter))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class PMUIStarterDelegateService : IPMUIStarter
```

Comment:
fwiw, while this service is `public`, it can only be called from within our code, so while I 100% agree we need to use good names/sensible signatures, it's not the same scrutiny as say a public API in NuGet.Commands.

Reply:
> Instead of creating this new public MEF component, can we just inline launching the PMUI through the command where we need it?

I think @martinrrm is missing a good description about the reasoning behind this type. 

If you reference types within NuGet.Tools.dll (where the PM UI starting is implemented) from within NuGet.SolutionRestoreManager, you'd load the NuGet.Tools assembly way too early and cause an RPS regression.

Reply:
> Is that because the MEF component cache is scoped?

We don't publish the assembly with the interfaces anywhere. We just build and insert.

> External code can still create an instance of this class directly and use it since it's public so we have to consider that as well.

Not sure why we would care about that. In theory, people could take *any* assembly, and use reflection to access any type, public or internal. 
FWIW, we've made *a lot* of public API changes from within NuGet.Clients assemblies. 
https://github.com/NuGet/NuGet.Client/blob/dev/docs/nuget-sdk.md contains a great summary of our policies. 

The rule of thumb as a contributor is that if you're not changing a Public.API file, you're good :)

Reply:
Great, thanks for the added insight! That makes sense.

Since we aren't giving this for public consumption, I feel more comfortable that we can add this now and make a breaking change later to support more detailed scenarios of launching the PMUI, as needed.

If we do go with this service, I'd recommend changing the name to something more meaningful. PackageManagerService or if we want it to be specific to just launching / showing the PMUI, something like PackageManagerLaunchService, etc.